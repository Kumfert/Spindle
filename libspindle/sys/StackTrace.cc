//
// StackTrace.cc
//
// $Id: StackTrace.cc,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//

#include "spindle/StackTrace.h"

#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

StackTrace::StackTrace() { }


StackTrace::~StackTrace() {
  if (stack.size() > 0 ) { 
    fprintf(stderr,"\nError: Stack not empty");
    dump(stderr);
  }
}

void
StackTrace::resize( size_t targetSize, int lineNumber, char* fileName, char* functionName) {
  if ( targetSize == stack.size() ) { 
    pop();
    return;
  }
  const char* fn = ( fileName==0 ) ? "unknown" : fileName;
  const char* ft = ( functionName==0) ? "unknown" : functionName;
  fprintf(stderr,"\nWARNING: StackSize %d is incorrect! %s in \"%s\" - %d",stack.size(),ft,fn,lineNumber );
  if ( targetSize < stack.size() ) { 
    // somehow we've forgotten to pop some members off of the stack
    fprintf(stderr,"\n         Probably forgot to pop %d members off the stack", stack.size() - targetSize);
    fprintf(stderr,"\n         Before modification...");
    dump( stderr );
    while( targetSize < stack.size() ) {
      pop();
    }
    fprintf(stderr,"\n         After modification...");
    dump( stderr );
  } else if ( targetSize > stack.size() ) {
    // Somehow, we've forgotten to push some members onto the stack
    fprintf(stderr,"\n         Probably forgot to push %d members onto the stack", stack.size() - targetSize);
    fprintf(stderr,"\n         Be sure to check %s and the functions it calls", functionName);
    fprintf(stderr,"\n         Before Modification...");
    dump(stderr);
    while( targetSize > stack.size() ) {
      push(lineNumber, fileName, "Dummy entries: Check for missings pushes");
    }
    fprintf(stderr,"\n        After Modification...");
    dump(stderr);
  }
  pop();
}
  

void
StackTrace::dump( FILE * stream ) const {
  for( StackType::const_reverse_iterator it = stack.rbegin(), stop = stack.rend();
       it != stop; ++ it ) {
    const char * fn = (( (*it).function == 0 ) ? "unknown" : (*it).function );
    const char * fl = (( (*it).file == 0 ) ? "unknown" : (*it).file );
    fprintf( stream, "\n  %s in \"%s\" - %d",fn, fl, (*it).line );
  }
}

void 
StackTrace::reset() {
  stack.erase( stack.begin(), stack.end() );
}

StackTrace& 
StackTrace::get() {
  static StackTrace st;
  return st;
}
