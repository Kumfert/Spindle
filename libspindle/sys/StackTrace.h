//
// StackTrace.h
//
// $Id: StackTrace.h,v 1.2 2000/02/18 01:32:01 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1997, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//
//

#ifndef SPINDLE_STACK_TRACE_H_
#define SPINDLE_STACK_TRACE_H_

#include <stdio.h>

#ifndef SPINDLE_H_
#include "spindle/spindle.h"
#endif

#ifdef REQUIRE_OLD_CXX_HEADER_SUFFIX
#include "vector.h"
#else
#include <vector>
using std::vector;
#endif

SPINDLE_BEGIN_NAMESPACE

struct StackInfo {
  int line;
  const char * file;
  const char * function;
  StackInfo( const int lineNumber, const char* fileName, const char* functionName) :
    line(lineNumber), file(fileName), function(functionName) { }
  StackInfo(): line(-1), file(0), function(0) { }
}; 

class SpindleSystem;

class StackTrace {
  friend SpindleSystem;
public:
  typedef vector< StackInfo > StackType;

private:
  StackType stack;
  StackTrace( const StackTrace& src );    // no implementation
  void operator=( const StackTrace& src); // no implementation
  StackTrace();

public:
  static StackTrace& get();
  ~StackTrace();
  void push( int lineNumber, const char* fileName="Unknown", const char* functionName="Unknown" );
  const StackInfo& top() const;
  void pop();
  int size() const;
  void resize( size_t targetSize, int lineNumber, char* fileName, char* functionName);
  void setLastLine( const int i );

  void dump( FILE * stream ) const;
  void reset();
};

inline void
StackTrace::push(int lineNumber, const char* fileName, const char* functionName ) {
  stack.push_back( StackInfo( lineNumber, fileName, functionName ) );
}

inline const StackInfo& 
StackTrace::top() const { 
  return stack.back();
}

inline void
StackTrace::pop() {
  if ( stack.size() == 0) { return; }
  stack.pop_back();
}

inline int
StackTrace::size() const { 
  return stack.size();
}

inline void
StackTrace::setLastLine( const int i ) {
  stack.back().line = i;
}

SPINDLE_END_NAMESPACE

#endif


