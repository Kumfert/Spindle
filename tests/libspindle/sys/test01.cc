//
// src/sys/drivers/tests/test01.cc
//
// $Id: test01.cc,v 1.2 2000/02/18 01:32:11 kumfert Exp $
//
//  Gary Kumfert, Old Dominion University
//  Copyright(c) 1998, Old Dominion University.  All rights reserved.
// 
//  Permission to use, copy, modify , distribute and sell this software and
//  its documentation for any purpose is hereby granted without fee, 
//  provided that the above copyright notice appear in all copies and
//  that both that copyright notice and this permission notice appear
//  in supporting documentation.  Old Dominion University makes no
//  representations about the suitability of this software for any 
//  purpose.  It is provided "as is" without express or implied warranty.
//
///////////////////////////////////////////////////////////////////////
//
//  Tests the stack trace
//

#include "spindle/StackTrace.h"

#define STACK StackTrace::get()
#define SpindleFunctionEnter \
        STACK.push(__LINE__,__FILE__,__FUNC__); \
	int SPINDLE_StackSize_local_variable = STACK.size(); 


#define SpindleFunctionCall \
        STACK.setLastLine( __LINE__ );

#define SpindleFunctionReturn( RTRN );\
	STACK.resize(SPINDLE_StackSize_local_variable,__LINE__,__FILE__,__FUNC__); \
	return (RTRN); 

#define SpindleFunctionReturnVoid \
	STACK.resize(SPINDLE_StackSize_local_variable,__LINE__,__FILE__,__FUNC__); \
	return; 

void f1();
void f2();
void f3();

int g1();
int g2();
int g3();

void h1();
void h2();
void h3();

#define __FUNC__ "main(argv,argc)"
#ifdef HAVE_NAMESPACES
using namespace SPINDLE_NAMESPACE;
#endif

int main() {
  SpindleFunctionEnter;

#ifdef SPINDLE__NO_INSTANCE_COUNT
  fprintf( stderr ,"WARNING: Instance counting disabled for this compilation platform\n");
#endif

  SpindleFunctionCall f1();

  SpindleFunctionCall g1();

  SpindleFunctionCall h1(); // intentionally forgot to pop one in h3

  // SpindleFunctionReturnVoid;  // intentionally forgot to pop end to test 
  ++SPINDLE_StackSize_local_variable; --SPINDLE_StackSize_local_variable;  // for when stack is destroyed. 
}

#undef __FUNC__
#define __FUNC__ "f1()"
void f1() {
  SpindleFunctionEnter;
  SpindleFunctionCall  f2();
  SpindleFunctionReturnVoid;
}


#undef __FUNC__
#define __FUNC__ "f2()"
void f2() {
  SpindleFunctionEnter;
  SpindleFunctionCall  f3();




  SpindleFunctionCall  f3();



  SpindleFunctionCall  f3();
  SpindleFunctionReturnVoid;
}


#undef __FUNC__
#define __FUNC__ "f3()"
void f3() {
  SpindleFunctionEnter;
  STACK.dump( stdout );
  SpindleFunctionReturnVoid;
}


#undef __FUNC__
#define __FUNC__ "g1()"
int g1() {
  SpindleFunctionEnter;
  SpindleFunctionCall  g2();
  SpindleFunctionReturn( 1 );
}


#undef __FUNC__
#define __FUNC__ "g2()"
int g2() {
  SpindleFunctionEnter;
  SpindleFunctionCall g3();
  SpindleFunctionReturn( 2 );
}


#undef __FUNC__
#define __FUNC__ "g3()"
int g3() {
  SpindleFunctionEnter;
  STACK.dump( stdout );
  SpindleFunctionReturn( 3 );
}


#undef __FUNC__
#define __FUNC__ "h1()"
void h1() {
  SpindleFunctionEnter;
  SpindleFunctionCall h2();
  SpindleFunctionReturnVoid;
}


#undef __FUNC__
#define __FUNC__ "h2()"
void h2() {
  SpindleFunctionEnter;
  SpindleFunctionCall h3();
  SpindleFunctionReturnVoid;
}


#undef __FUNC__
#define __FUNC__ "h3()"
void h3() {
  SpindleFunctionEnter;
  STACK.dump( stdout );
  // SpindleFunctionReturnVoid; // intentionally forgot this one
  ++SPINDLE_StackSize_local_variable; --SPINDLE_StackSize_local_variable;  // for when stack is destroyed. 
  // keeps some compilers from complaining 
  // would be used if I didn't intentionally forget
				   
}

