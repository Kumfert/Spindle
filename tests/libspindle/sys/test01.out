
WARNING: StackSize 4 is incorrect! h2() in "test01.cc" - 152
         Probably forgot to pop 1 members off the stack
         Before modification...
  h3() in "test01.cc" - 159
  h2() in "test01.cc" - 151
  h1() in "test01.cc" - 142
  main(argv,argc) in "test01.cc" - 69
         After modification...
  h2() in "test01.cc" - 151
  h1() in "test01.cc" - 142
  main(argv,argc) in "test01.cc" - 69
Error: Stack not empty
  main(argv,argc) in "test01.cc" - 69
  f3() in "test01.cc" - 105
  f2() in "test01.cc" - 88
  f1() in "test01.cc" - 79
  main(argv,argc) in "test01.cc" - 65
  f3() in "test01.cc" - 105
  f2() in "test01.cc" - 93
  f1() in "test01.cc" - 79
  main(argv,argc) in "test01.cc" - 65
  f3() in "test01.cc" - 105
  f2() in "test01.cc" - 97
  f1() in "test01.cc" - 79
  main(argv,argc) in "test01.cc" - 65
  g3() in "test01.cc" - 132
  g2() in "test01.cc" - 124
  g1() in "test01.cc" - 115
  main(argv,argc) in "test01.cc" - 67
  h3() in "test01.cc" - 159
  h2() in "test01.cc" - 151
  h1() in "test01.cc" - 142
  main(argv,argc) in "test01.cc" - 69